Render VmDetails with deleted template curreclty.

yaacov opened the pull request:
**Description**
When we want to render VmDetails with missing VmTemplate we get a warning that we can not find the template. But It is OK to delete a template of an existing vm, the view should silently fall back to vm's current resources.

Ref: https://bugzilla.redhat.com/show_bug.cgi?id=1685393

**Screenshots**

When template is missing, show warning triangle instead of an error:

![Peek 2019-03-31 14-35](https://user-images.githubusercontent.com/2181522/55288681-ac8a8c80-53c3-11e9-8381-5e0517b06073.gif)

Replace drop-down with a label when only one flavor option is available:

![Peek 2019-03-31 16-34](https://user-images.githubusercontent.com/2181522/55289927-49095a80-53d5-11e9-8ec2-95095e2ef76f.gif)

Add an optional template resource link method to vmDetails:

![Peek 2019-04-01 11-43](https://user-images.githubusercontent.com/2181522/55314707-64c33e00-5473-11e9-9dcb-eadc61c62fc6.gif)

Show labels for CPU and Memory in flavor form:

![Peek 2019-04-01 14-46](https://user-images.githubusercontent.com/2181522/55325218-0acf7200-548d-11e9-91b1-7d7aa46cd874.gif)

yaacov commented:
wip until fixing coverage ...

suomiy commented on src/components/Details/VmDetails/VmDetails.js:
```diff
@@ -123,10 +129,17 @@ export class VmDetails extends React.Component {
+
+    // Check template, if template is missing, update view.
+    if (getVmTemplate(vm) != null && this.state.templateError === PENDING_STR) {
+      retrieveVmTemplate(k8sGet, vm)
```
retrieveVmTemplate is already called for the flavor. Please call this method only once.

yaacov replied:
:+1:

yaacov commented on src/components/Details/VmDetails/VmDetails.js:
```diff
@@ -123,10 +129,17 @@ export class VmDetails extends React.Component {
     }
+
+    // Check template, if template is missing, update view.
+    if (getVmTemplate(vm) != null && this.state.templateError === PENDING_STR) {
```
Thanks ! will fix that.

rawagner commented on src/components/Details/VmDetails/tests/VmDetails.test.js:
```diff
@@ -15,6 +15,8 @@ import { getId } from '../../../../selectors';
 
 const testVmDetails = (vm, otherProps) => <VmDetails {...VmDetailsFixture[0].props} vm={vm} {...otherProps} />;
 
+const flushAllPromises = () => new Promise(resolve => setImmediate(resolve));
```
you can reuse https://github.com/kubevirt/web-ui-components/blob/master/src/tests/enzyme.js#L3

yaacov replied:
Nice, thanks :exclamation: :+1:

yaacov commented:
@rawagner @suomiy please re-review.

I'v removed the part that indicate missing template in the vm details view, because it is not part of the BZ, and may be implemented better in a separate PR.

rawagner commented:
@yaacov I wonder if its okay to not show anything in UI if we cannot load the template ? We use the loaded template to get possible flavors when editing. Lets consider this flow:
1. I will create VM from template with Small flavor
2. I will delete the template (or somebody else)
3. I will go to VM details - I can see that flavor is Small
4. I will go to Edit mode - flavor dropdown now shows only Custom Flavor and I might get confused why is that so.

@matthewcarleton WDYT ?

suomiy commented on src/components/Details/Flavor/Flavor.js:
```diff
@@ -44,13 +44,14 @@ export class Flavor extends React.Component {
-        return this.setState({
+      .catch(error =>
+        // An error occurred while loading template, it's ok, template may be deleted.
+        // Ref: BZ1685393
```
is it necessary to include the BZ number?

yaacov replied:
@rawagner , 

> is it necessary to include the BZ number?

No :-) , but it is strange not to do anything on error ... so I needed some excuse.
looking again at @rawagner comment https://github.com/kubevirt/web-ui-components/pull/287#issuecomment-476177652 it may be better to find some way to indicate that the template is missing ... so this will be replaced with actually doing something, and no need for the BZ ref.

yaacov replied:
@suomiy sorry ^^

yaacov commented:
@rawagner 
> I wonder if its okay to not show anything in UI if we cannot load the template ?

I agree, will look for a way to indicate missing template to the user.

yaacov commented:
@rawagner @suomiy please re-review

Changes:
moved  `retrieveVmTemplate` one computnent higher from falvor to vm - 
pros:
- knowledge about actual template in etcd is now in the vm sate, and can be used in rendering vm details.
 - vmTemplate that get the template (with it's flavors) as a prop do need to re-read it from etcd when editing it's flavor.
- added tests for vm details with deleted template, and flavor in edit state.

suomiy commented on src/components/Details/VmTemplateDetails/VmTemplateDetails.js:
```diff
@@ -218,9 +212,8 @@ export class VmTemplateDetails extends React.Component {
-                      retrieveVmTemplate={() => retrieveVmTemplate(k8sGet, vmTemplate)}
                       formValues={settingsValue(this.state.form, FLAVOR_KEY)}
-                      onLoadError={this.onLoadError}
+                      template={vmTemplate}
```
this component should still retrieve the base template for flavor options

yaacov replied:
@suomiy , can we have a recursion of template depending on another template ?

will re-add loading of base template, just to be sure, but not sure if we need to, WDYT ?

suomiy replied:
not real recursion (just the recursion of types (: ) But yes, user vm templates are depending on pre-instaled templates (that's where we get our flavors from).

yaacov replied:
:+1:  Thanks !

suomiy commented:
> vmTemplate that get the template (with it's flavors) as a prop do need to re-read it from etcd when editing it's flavor.

not sure this is actually happening

yaacov commented:
> not sure this is actually happening

Sorry, typo, I meant to write "prop ~do~ **do not** need to re-read it from etcd when editing it's flavo"

yaacov commented on src/components/Details/VmDetails/VmDetails.js:
```diff
@@ -122,7 +124,33 @@ export class VmDetails extends React.Component {
+      )
+      .catch(error =>
+        this.setState({
+          updating: false,
```
make sense to me :+1:

yaacov commented on src/components/Details/VmDetails/VmDetails.js:
```diff
@@ -122,7 +124,33 @@ export class VmDetails extends React.Component {
+    const { k8sGet, vm } = this.props;
+
+    // if already loading or have template, just return.
+    if (this.state.template || this.state.updating) {
```
Right :clap: removing

yaacov commented on src/components/Details/Flavor/fixtures/Flavor.fixture.js:
```diff
@@ -37,24 +39,19 @@ export default [
-        new Promise(resolve => {
-          resolve(fedora28);
-        }),
+      template: cloudInitTestVm,
```
@suomiy 
I think it should be `fedora28` ?
like it was on old line 56 ?

suomiy replied:
I mean, it doesn't need a template when in updating state? Nevertheless both should be fine.

yaacov replied:
:+1:

yaacov commented on src/components/Details/Flavor/Flavor.js:
```diff
@@ -31,32 +27,20 @@ export class Flavor extends React.Component {
+  componentDidUpdate(prevProps, prevState) {
+    if (this.props.template && prevProps.template !== this.props.template) {
+      this.props.onFormChange(
+        this.props.template,
```
Oh... this should be done on the vm side ... will remove `componentDidUpdate` from flavor.

matthewcarleton commented:
> @yaacov I wonder if its okay to not show anything in UI if we cannot load the template ? We use the loaded template to get possible flavors when editing. Lets consider this flow:
> 
> 1. I will create VM from template with Small flavor
> 2. I will delete the template (or somebody else)
> 3. I will go to VM details - I can see that flavor is Small
> 4. I will go to Edit mode - flavor dropdown now shows only Custom Flavor and I might get confused why is that so.
> 
> @matthewcarleton WDYT ?

If I'm understanding this correctly, I agree that we shouldn't show an error if the template is deleted. I think we should show under templates that the template no longer exists. We might need to explore this in the design a bit further. I would say if all we can offer is a custom flavour at that point then we should not offer a dropdown with just one option. It could just show as a heading of "custom" with the fields for CPUs and Memory. Is the idea that the values of the small flavour will now be mapped to a custom set of values?

yaacov commented:
> think we should show under templates that the template no longer exists. We might need to explore this in the design a bit further.

:+1: 

> I would say if all we can offer is a custom flavour at that point then we should not offer a dropdown with just one option.

Make sense to me, this may be a more global issue, how should we render any dropdown with only one option, IMHO this is also an issue for more exploring.

> Is the idea that the values of the small flavour will now be mapped to a custom set of values?

IFAIK no, IIUC the kubevirt operator running on the server will continue to use values from the now missing template, until we actively  push a change.

suomiy commented:
> Make sense to me, this may be a more global issue, how should we render any dropdown with only one option, IMHO this is also an issue for more exploring.

we can start with the flavor and then resolve other use cases in the future

yaacov commented:
@suomiy the testing looks stuck :-( 

> should show under templates that the template no longer exists

a "pficon-warning-triangle-o" with tooltip , WDYT ?

>  how should we render any dropdown with only one option,

make it disabled, WDYT ?

can you re-review current state of code ?

suomiy commented on src/components/Details/VmDetails/fixtures/VmDetails.fixture.js:
```diff
@@ -87,6 +87,36 @@ export const vmFixtures = {
+      },
+      running: false,
+    },
+    editing: true,
```
should this line be included in the vm fixture?

suomiy commented on src/components/Details/common/tests/details.js:
```diff
@@ -58,6 +59,7 @@ export const disablesEditOnCancel = component =>
   awaitVmDetails(() => {
     expect(component.find(InlineFormFactory).exists()).toBeFalsy();
     clickButton(component, 'Edit');
+    component.setState({ updating: false });
```
is this because of the waiting on retrieveVmTemplate? Can we rather mock it and wait for the promise to finish?

yaacov replied:
> Can we rather mock it and wait for the promise to finish?

I can, It's already mocked, I tried to wait for the promise and then removed the `async` and `await flushPromises` because I couldn't make it look nice, I can try again, maybe re-factor it this time to look nicer :-)

suomiy commented:
> @suomiy the testing looks stuck :-(

not sure what is going on, maybe the next push will fix it :)

> > should show under templates that the template no longer exists

> a "pficon-warning-triangle-o" with tooltip , WDYT ?

@matthewcarleton ?
sounds good to me, we could also implement it in this PR 

> > how should we render any dropdown with only one option,

> make it disabled, WDYT ?

> It could just show as a heading of "custom" with the fields for CPUs and Memory.

sounds good, it doesn't have to be a dropdown,

matthewcarleton commented:
@suomiy, @yaacov 
This is what I'm thinking. Would we want to show the action related to the VM no longer being available? "This template has been deleted" ? My only question is, if the user is looking to recover that template is that possible? 
<img width="604" alt="Screen Shot 2019-03-27 at 2 14 21 PM" src="https://user-images.githubusercontent.com/287880/55101574-c57d0000-509a-11e9-9cce-0d234877110b.png">
I like the idea of having just a heading for flavour rather than a disabled dropdown. A disabled dropdown doesn't indicate why it's disabled so it could be confusing.

yaacov commented:
> My only question is, if the user is looking to recover that template is that possible?

AFAIK no, it is not possible to recover deleted user defined templates that user deleted.

matthewcarleton commented:
@yaacov ok, if that's the case then we wouldn't need any action in the popover. If it were, I was thinking we could offer a recover option.

yaacov commented:
@mareklibra @suomiy does it make sense to make the template a link, if it exist ?

matthewcarleton commented:
@yaacov I don't think it's been shown in the design, but I would assume if the template is available it would be a link.

suomiy commented:
> does it make sense to make the template a link, if it exist ?

I guess yes, for the vmDetail. The template detail not so much, because even though the template exists, the representation of that object  in oc console does not (no generic templates page).

yaacov commented:
Added a warning triangle when template is missing:

![Peek 2019-03-31 14-35](https://user-images.githubusercontent.com/2181522/55288681-ac8a8c80-53c3-11e9-8381-5e0517b06073.gif)

yaacov commented:
Replace drop-down with a label when only one option is available:

![Peek 2019-03-31 16-34](https://user-images.githubusercontent.com/2181522/55289927-49095a80-53d5-11e9-8ec2-95095e2ef76f.gif)

yaacov commented:
Add an optional template resource link method to vmDetails:

![Peek 2019-04-01 11-43](https://user-images.githubusercontent.com/2181522/55314707-64c33e00-5473-11e9-9dcb-eadc61c62fc6.gif)

suomiy commented on src/components/Details/VmTemplateDetails/VmTemplateDetails.js:
```diff
@@ -124,10 +142,10 @@ export class VmTemplateDetails extends React.Component {
     const id = getId(vmTemplate);
     const vm = selectVm(vmTemplate.objects);
-    const baseTemplate = getVmTemplate(vmTemplate);
+    const baseTemplateLabels = getVmTemplate(vmTemplate);
```
please can you rename this method to something like getVmTemplateLabels/getVmTemplateInfo. Or even better rename it to getTemplateDisplayName and reimplement it so that it returns string?

yaacov replied:
sure :smile_cat:

suomiy commented:
can we add some margin to the custom label so it appears similar to the dropdown ? Not sure if that is necessary, just my opinion. 

![mar](https://user-images.githubusercontent.com/3648838/55319117-8244d780-5474-11e9-89b7-1b37bfb9fb75.png)

suomiy commented on src/components/Details/VmTemplateDetails/VmTemplateDetails.js:
```diff
@@ -145,7 +145,10 @@ export class VmTemplateDetails extends React.Component {
     const id = getId(vmTemplate);
     const vm = selectVm(vmTemplate.objects);
-    const baseTemplateLabels = getVmTemplate(vmTemplate);
+    const getTemplateDisplayName = () => {
```
I meant, can we replace the getVmTemplate method? So it can be also reused in vmDetails.

yaacov replied:
:-)

yaacov replied:
I think it's ok now :white_flower:

yaacov commented:
> can we add some margin to the custom label so it appears similar to the dropdown ?

```
.kubevirt-flavor__label {
   margin-bottom ...
}
```
@suomiy  WDYT ?

yaacov commented:
@suomiy @matthewcarleton do we need some labels to the inputs in edit mode ?

![Peek 2019-04-01 11-43](https://user-images.githubusercontent.com/3648838/55319117-8244d780-5474-11e9-89b7-1b37bfb9fb75.png)

we curently have:
```
[ flavor drop down ]
[ number input     ]
[ number input     ]
```
should it have labels in edit mode, e.g.
```
[ flavor drop down        ]
Memory [ number input     ]
CPU    [ number input     ]
```
WDYT ?

b - Does `<InlineEdit formFields` has some props for that ?

suomiy commented:
> ```
> .kubevirt-flavor__label {
>   margin-bottom ...
> }
> ```
> @suomiy WDYT ?

yup, I would make the margin similar to the dropdown

> [ flavor drop down        ]
> Memory [ number input     ]
> CPU    [ number input     ]
>
> WDYT ?

> b - Does <InlineEdit formFields has some props for that ?

+1, that is a good remark. I guess it has to be some regression with InlineFormFactory, because It already has a title. Can you look at it?

yaacov commented:
> I guess it has to be some regression with InlineFormFactory, because It already has a title. Can you look at it?

Sure :+1:

suomiy commented on src/components/Details/VmTemplateDetails/VmTemplateDetails.js:
```diff
@@ -124,10 +143,10 @@ export class VmTemplateDetails extends React.Component {
     const id = getId(vmTemplate);
     const vm = selectVm(vmTemplate.objects);
-    const baseTemplate = getVmTemplate(vmTemplate);
+    const baseTemplateLabels = getTemplateDisplayName(getVmTemplate(vmTemplate));
```
nit: variable name

yaacov replied:
sorry ... `baseTemplateName` ?

suomiy replied:
np, yup could be or baseTemplateDisplayName

yaacov commented:
show labels in inline form:

![Peek 2019-04-01 14-46](https://user-images.githubusercontent.com/2181522/55325218-0acf7200-548d-11e9-91b1-7d7aa46cd874.gif)

yaacov commented:
@suomiy this is nearing the end :tada:  :balloon: , should I squash, or leave separate commits for easier review ?

matthewcarleton commented:
@yaacov this looks great, thanks! One question, what happens when the available CPU/Memory is exceeded?

yaacov commented:
> what happens when the available CPU/Memory is exceeded?

@matthewcarleton 

TL;DR - I do not know.

IFAIK we do not have the `quota/total` values to validate the form ... @suomiy do you know something about this ? 

p.s.
What I think happen when values are bad is that openshift will try to execute and fail after some tries ... @suomiy - do you know something about that ?

yaacov commented:
@suomiy @rawagner  please review ... I think it's ready :beers:

suomiy commented:
> > > what happens when the available CPU/Memory is exceeded?
> 
> @matthewcarleton
> 
> TL;DR - I do not know.
> 
> IFAIK we do not have the quota/total values to validate the form ... @suomiy do you know > something about this ?
> 
> p.s.
> What I think happen when values are bad is that openshift will try to execute and fail after >some tries ... @suomiy - do you know something about that ?

Currently, we do not have a way to know the quota AFAIK. The request will go ok, just the VM will fail with an error the next time you try to start it.

suomiy commented:
> @suomiy this is nearing the end tada  balloon , should I squash, or leave separate commits for easier review ?

don't worry about that. It will be squashed on merge

suomiy commented:
> .kubevirt-flavor__label {
>   margin-bottom ...
> } 

is this present in this PR?

yaacov commented:
>    .kubevirt-flavor__label {
>    margin-bottom ...
>    }
>
> is this present in this PR?

Aaa... forgot to do `git add` :100:

matthewcarleton commented:
@suomiy ah ok. Ya it would be a nicer experience for sure to know what the limit is at the time. Would it be possible to explore adding that feature (in another PR :) )

suomiy commented:
> Would it be possible to explore adding that feature (in another PR :) )

I think this should be explored on the backend. Doesn't seem like we can do much about it here in the UI.

@jelkosz Do you have some additional information on this issue?

suomiy commented:
looks good, merging
